net/captivedetection/endpoints.go

```diff
@@ -0,0 +1,63 @@
+// Copyright (c) Tailscale Inc & AUTHORS
+// SPDX-License-Identifier: BSD-3-Clause
+
+package captivedetection
+
+import (
+	"net/http"
+	"runtime"
+	"slices"
+
+	"tailscale.com/net/dnsfallback"
+)
+
+// availableEndpoints returns a set of Endpoints which can be used for captive portal detection by performing
+// one or more HTTP requests and looking at the response. The returned Endpoints are ordered by preference,
+// with the most preferred Endpoint being the first in the slice.
+func availableEndpoints(preferredDERPRegionID int) []Endpoint {
+	endpoints := []Endpoint{}
+
+	// Before anything, if we have a preferred DERP region, let's try to use that as our first endpoint.
+	preferredIPsAndHosts := []string{}
+	if preferredDERPRegionID != 0 {
+		hs := dnsfallback.GetDERPHostnames(preferredDERPRegionID)
+		ips := dnsfallback.GetDERPIPv4s(preferredDERPRegionID)
+		preferredIPsAndHosts = slices.Concat(ips, hs)
+	}
+
+	// First, try to use the DERP servers. Generate URLs for both their IPs and hostnames (IPs first).
+	// These might be blocked on some networks since they are IPs and hostnames associated with Tailscale.
+	derps := slices.Concat(preferredIPsAndHosts, dnsfallback.GetAllDERPIPv4s(), dnsfallback.GetAllDERPHostnames())
+	for _, derp := range derps {
+		url := "http://" + derp + "/generate_204"
+		e := Endpoint{url, http.StatusNoContent, "", true}
+		endpoints = append(endpoints, e)
+	}
```

Comment:
We have a lot of DERP regions; I think this means that we'll always have >6 endpoints and thus never select the well-known stuff below?

Reply:
Yeah, this needs fixing. I'll change it to make sure we always choose both a DERP AND a platform-specific server

Reply:
I wonder if rather than a platform-specific server, we can see if there's a platform-specific API that will tell us what the OS's probing of the platform-specific server concluded.

Reply:
As far as Apple platforms are concerned, this is not possible on iOS. It is possible on macOS for now, but the API (`SystemConfiguration/CaptiveNetwork.h`) is being deprecated and will likely disappear soon.